lib/Apache2_4/AuthCookieDBI.pm

```diff
@@ -0,0 +1,225 @@
+# Apache2_4::AuthCookieDBI
+#
+# A module implementing Apache 2.4.x compatibility for Apache2::AuthCookieDBI
+# group-based authorizations.
+#
+# See end of this file for Copyright notices.
+#
+# Maintainer: Matisse Enzer <matisse@cpan.org> (as of version 2.0)
+#
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+#
+#===============================================================================
+
+package Apache2_4::AuthCookieDBI;
+
+# ABSTRACT: Perl group authorization for Apache 2.4.x
+
+use strict;
+use base 'Apache2::AuthCookieDBI';
+use Apache2::Log;
+use Apache2::Const -compile => qw(AUTHZ_GRANTED AUTHZ_DENIED AUTHZ_DENIED_NO_USER AUTHZ_GENERAL_ERROR);
+use Apache::AuthCookie::Util qw(is_blank);
+
+#===============================================================================
+# FILE (LEXICAL)  G L O B A L S
+#===============================================================================
+
+our $VERSION = '2.19';
+use constant LOG_TYPE_AUTHZ   => 'authorization';
+
+#===============================================================================
+# M E T H O D ( S )
+#===============================================================================
+
+sub group {
+    my ($class, $r, $groups) = @_;
+
+    my $debug = $r->dir_config("AuthCookieDebug") || 0;
+
+    my $user = $r->user;
+
+    $r->server->log_error("authz:start user=@{[ defined($user) ? $user : '(undef)' ]} type=$class groups=@{[ defined($groups) ? $groups : '(undef)' ]} uri=@{[ $r->uri ]}") if ($debug >= 5);
+
+    if (is_blank($user)) {
+        # User is not yet authenticated.
+        return Apache2::Const::AUTHZ_DENIED_NO_USER;
+    }
+
+    if (is_blank($groups)) {
+        $r->server->log_error("${class}\tno group(s) specified in \'Require group ...\' configuration for URI @{[ $r->uri ]}");
+        return Apache2::Const::AUTHZ_DENIED;
+    }
+
+    my @groups = split( Apache2::AuthCookieDBI::WHITESPACE_REGEX, $groups );
+
+    # Instead of querying the database every time, check subprocess_env
+    # for AUTH_COOKIE_DBI_GROUP (set below) and utilize that to short circuit
+    # the authorization.
+    if ( my $cached_group = $r->subprocess_env('AUTH_COOKIE_DBI_GROUP') ) {
+	$r->server->log_error("${class}\tfound cached group $cached_group") if ($debug >= 3);
+	foreach my $group (@groups) {
```

Comment:
I think we can skip the foreach loop here, and move this check up to the top of the method, potentially returning even earlier.

My thinking is that if `AUTH_COOKIE_DBI_GROUP` subprocess_env is set then we have already returned `AUTHZ_GRANTED` at least once, and so we might as well keep returning `AUTHZ_GRANTED` as early as possible in this method.

Does that make sense?